Azure updates (#607)
strapi/documentation: docs/developer-docs/latest/setup-deployment-guides/deployment/hosting-guides/azure.md

---

# Azure
!!!include(developer-docs/latest/setup-deployment-guides/deployment/snippets/deployment-guide-not-updated.md)!!!

> This is a step-by-step guide for deploying a Strapi project to [Azure](https://azure.microsoft.com/) using Platform-as-a-Service (PaaS). If you're interested in using Infrastructre-as-a-Service (IaaS) refer to [IaaS deployment guide](#iaas-deployment-guide) further down. Prior to starting this guide, you should have created a [Strapi project](/developer-docs/latest/getting-started/quick-start.md) and have read through the [configuration](/developer-docs/latest/setup-deployment-guides/deployment.md#application-configuration) section.

Reviewer: ```suggestion
This is a step-by-step guide for deploying a Strapi project to [Azure](https://azure.microsoft.com/) using Platform-as-a-Service (PaaS). If you're interested in using Infrastructure-as-a-Service (IaaS) refer to [IaaS deployment guide](#iaas-deployment-guide).

::: prerequisites
* You should have created a [Strapi project](/developer-docs/latest/getting-started/quick-start.md)/
* You have should read through the [configuration documentation](/developer-docs/latest/setup-deployment-guides/deployment.md#application-configuration).
:::
```

> ## Azure Install Requirements

Reviewer: ```suggestion

```

> - You must have an [Azure account](https://azure.microsoft.com/free/) before doing these steps.

Reviewer: Please include this in the `::: prerequisites` callout (see comment above).

## PaaS Deployment Guides

> There are three ways which you can deploy the Azure resources, via the [Azure Portal](#creating-resources-via-the-azure-portal), via the [Azure CLI](#creating-resources-via-the-azure-cli) or via an [Azure Resource Manager template](#deploy-with-an-azure-resource-manager-template).

Reviewer: ```suggestion
There are 3 ways which you can deploy the Azure resources:
- via the [Azure Portal](#creating-resources-via-the-azure-portal),
- via the [Azure CLI](#creating-resources-via-the-azure-cli),
- or via an [Azure Resource Manager template](#deploy-with-an-azure-resource-manager-template).
```

When Strapi is running in a PaaS hosting model, a custom storage provider will be required to avoid the transient disk of the PaaS model, [which is covered towards the end](#configure-strapi-for-azure-appservice).

### Required Resources

There are three resources in Azure that are required to run Strapi in a PaaS model, [AppService](https://azure.microsoft.com/services/app-service/?WT.mc_id=javascript-37811-aapowell) to host the Strapi web application, [Storage](https://azure.microsoft.com/product-categories/storage/?WT.mc_id=javascript-37811-aapowell) to store images/uploaded assets, and a database, Azure has managed MySQL and Postgres to choose from (for this tutorial, we'll use MySQL, but the steps are the same for MySQL).

### Creating Resources via the Azure Portal

In this section we'll use the Azure Portal to create the required resources to host Strapi.

1. Navigate to the [Azure Portal](https://portal.azure.com/?WT.mc_id=javascript-37811-aapowell)

> 1. Click **Create a resource** and search for _Resource group_ from the provided search box

Reviewer 2: All of your steps have the same number

Author: Numbers for an ordered list auto-increment and don't require explicit stating (unless you need to restart lists due to a break introduced), see the published build here https://documentation-k9ok0crap-strapijs.vercel.app/developer-docs/latest/setup-deployment-guides/deployment/hosting-guides/azure.html#creating-resources-via-the-azure-portal and the numbers increment as expected.

By using the same start number, it's easier to introduce new steps, reorder steps, etc. without bulk changes throughout the doc.

The original doc (on deploying to a VM) didn't use ordered lists, it uses numbered headings (which I find breaks the anchor links), which is why they were explicitly stated.

Reading the [writing guide](https://handbook.strapi.io/user-success-manual/strapi-documentation-style-guide#1d4a11f683d047cd9128664823db2b7d) it didn't seem to state one way or the other, but if the intent is to have the Markdown explicitly reflect the steps, I'll update to reflect.

Reviewer: I agree it works with our VuePress setup but it still makes things easier to review/debug/discuss with colleagues. So when you have time, it would be nice if you could please update these numbers.

> 1. Provide a name for your Resource Group, `my-strapi-app`, and select a region

Reviewer: Missing dots at the end of each sentence, could you please add them?

1. Enter a username and password for the _Administrator account_, click **Review + create** then **Create**

#### Configuring the Resources

Once all the resources are created, you will need to get the connection information for the MySQL and Storage account to the Web App, as well as configure the resources for use.

##### Configure the Storage Account

1. Navigate to the Storage Account resource, then **Data storage** - **Containers**
1. Create a new Container, provide a _Name_, `strapi-uploads`, and set _Public access level_ to `Blob`, then click **Create**
1. Navigate to **Security + networking** - **Access keys**, copy the _Storage account name_ and _key1_
1. Navigate to the **Web App** you created and go to **Settings** - **Configuration**
1. Create new application settings for the Storage account, storage account key and container name (these will become the environment variables available to Strapi) and click _Save_

##### Configure MySQL

1. Navigate to the MySQL resource then **Settings** - **Connection security**
1. Set `Allow access to Azure services` to `Yes` and click **Save**
1. Navigate to **Overview** and copy _Server name_ and _Server admin login name_

> 1. Open the [Azure Cloud Shell](https://shell.azure.com?WT.mc_id=javascript-37811-aapowell) and log into the `mysql` cli:

Reviewer: ```suggestion
1. Open the [Azure Cloud Shell](https://shell.azure.com?WT.mc_id=javascript-37811-aapowell) and log into the `mysql` CLI:
```

   - `mysql --host <server> --user <username> -p`

1. Create a database for Strapi to use `CREATE DATABASE strapi;` then close the Cloud Shell

>    - Optional - create a separate non server admin user (see [this doc](https://docs.microsoft.com/azure/mysql/howto-create-users?tabs=single-server&WT.mc_id=javascript-37811-aapowell) for guidance)

Reviewer: ```suggestion
   - (_Optional_) - Create a separate non-server admin user (see [Azure documentation](https://docs.microsoft.com/azure/mysql/howto-create-users?tabs=single-server&WT.mc_id=javascript-37811-aapowell) for guidance)
```

   ```

### Deploy with an Azure Resource Manager template

To deploy using an Azure Resource Manager template, use the botton below, or upload [this template](https://gist.githubusercontent.com/aaronpowell/f216f119ffe5ed52945b46d0bb55569b/raw/2490a9295ea25c905096dbaae57da6ef8edb0e43/azuredeploy.json) as a custom deployment in Azure.

[![Deploy to Azure](https://aka.ms/deploytoazurebutton)](https://portal.azure.com/#create/Microsoft.Template/uri/https%3A%2F%2Fgist.githubusercontent.com%2Faaronpowell%2Ff216f119ffe5ed52945b46d0bb55569b%2Fraw%2F2490a9295ea25c905096dbaae57da6ef8edb0e43%2Fazuredeploy.json)

### Configure Strapi for Azure AppService

Azure AppService can be deployed to using CI/CD pipelines or via FTPS, refer to the [Azure docs](https://docs.microsoft.com/azure/app-service/deploy-continuous-deployment?tabs=github&WT.mc_id=javascript-37811-aapowell) on how to do this for your preferred manner.

As AppService is a PaaS hosting model, an upload provider will be required to save the uploaded assets to Azure Storage. Check out https://github.com/jakeFeldman/strapi-provider-upload-azure-storage for more details on using Azure Storage as an upload provider.

> _Note - for local development, you can either use the local disk upload provider, or the Azure Storage upload provider against the [Storage emulator](https://docs.microsoft.com/azure/storage/common/storage-use-azurite?toc=%2Fazure%2Fstorage%2Fblobs%2Ftoc.json&tabs=visual-studio&WT.mc_id=javascript-37811-aapowell)._

Reviewer: ```suggestion
::: note

For local development, you can either use the local disk upload provider, or the Azure Storage upload provider against the [Storage emulator](https://docs.microsoft.com/azure/storage/common/storage-use-azurite?toc=%2Fazure%2Fstorage%2Fblobs%2Ftoc.json&tabs=visual-studio&WT.mc_id=javascript-37811-aapowell).
:::
```

To start the Node.js application, AppService will run the `npm start` command. As there is no guarantee that the symlinks created by `npm install` were preserved (in the case of an upload from a CI/CD pipeline) it is recommended that the `npm start` command directly references the Strapi entry point:

```json
"scripts": {
    "start": "node node_modules/strapi/bin/strapi.js start"
}
```

## IaaS Deployment Guide

Databases can be on a [Azure Virtual Machine](https://azure.microsoft.com/services/virtual-machines/), hosted externally as a service, or via the [Azure Managed Databases](https://azure.microsoft.com/services/postgresql/).

> ### Azure Install Requirements

Reviewer: Please convert this to a `::: prerequisites` callout, using bullet points.
